main_code/Case_1/warehouse_reminder.py

```diff
@@ -0,0 +1,282 @@
+import pandas as pd
+import datetime
+from dateutil import parser
+import smtplib
+from email.mime.text import MIMEText
+from email.mime.multipart import MIMEMultipart
+import os
+import logging
+
+###################
+# global vars
+###################
+
+SPREADSHEET_FILE_LOCATION = 'ex_1.xlsx'
+SEND_EMAILS_FROM = 'mccul157@msu.edu'
+DEBUG_VAR = True
+
+
+###################
+# classes
+###################
+
+
+class Part(object):
+
+    def __init__(self, part_id, date, amount, open, contact, from_email, server) -> None:
+        self.id = part_id
+        self.date = self.convert_to_date_obj(date)
+        self.amount = amount
+        self.open = open
+        self.email = contact
+        self.from_email = from_email
+        self.server = server
+
+    def __repr__(self):
+        return('ID: {id}, Due: {date}, Recieved: {amount}, '
+              'In-Shipping: {open}, Contact: {email}, Send From: {from_email}'.format(
+               id=self.id, date=self.date, amount=self.amount, open=self.open, email=self.email,
+               from_email=self.from_email
+        ))
+
+    def check_if_action_required(self) -> bool:
+        """
+        Determines if action is required based on date and value of open and amount
+        :return: bool
+        """
+
+        current_time = datetime.datetime.today()
+
+        # we store our date as an object called date, but datetime uses objects called datetime
+        # so we use .combine our date object with the minimal time interval to get the right
+        # kind of object
+        if datetime.datetime.combine(self.date,
+                                     datetime.datetime.min.time()) - datetime.timedelta(days=5) <= current_time:
+            return True
+        return False
+
+    def send_email(self) -> None:
+        """
+        Sends email with the part ID, how many ordered, how many still accepting as reminder
+        :return: None
+        """
+
+        # just like earlier we use this to convert date into datetime
+        days_between_deadline_and_now = datetime.datetime.combine(
+            self.date, datetime.datetime.min.time()) - datetime.datetime.today()
+
+        static_email_text = '''
+        Hello,
+        \tThis is an automated email reminder about ordering part {part_id}.  At this time only {amount}
+        have been recieved and {open} are currently on route.  As of now there are less than {days} left
+        before the deadline on {date} has passed.
+        '''.format(open=self.open, part_id=self.id, amount=self.amount, date=str(self.date),
+                   days=int(days_between_deadline_and_now.days))
+
+        msg = MIMEMultipart()
+        msg['Subject'] = 'AUTOMATED Part Order Reminder'
+        msg['From'] = self.from_email
+        msg['To'] = self.email
+        # msg.attach(static_email_text)
+        body = MIMEText(static_email_text)
+        msg.attach(body)
+
+        self.server.sendmail(self.from_email, self.email, msg.as_string())
```

Comment:
self.server.sendmail(self.from_email, self.email, msg.as_string())

I am confused about this line.

Reply:
So we have our locally start email server, and server is something we have passed to our part class, (server is a singleton in this case kinda). This is taking the constructed email body msg and then using the server to actually send the email.  self.from_email represents what email we are sending it from, and self.email represents the target email address of where we are sending it.